Source/santabs/SNTBundleService.m

```diff
@@ -0,0 +1,275 @@
+  __block uint64_t binaryCount = 0;
+  __block uint64_t sentBinaryCount = 0;
+  __block uint64_t fileCount = 0;
+
+  __block BOOL breakDir = NO;
+
+  // In the first stage iterate over every file in the tree checking if it is a binary. If so add
+  // it to the fis set for the second stage. Hashing the file while iterating over the filesystem
+  // causes some weird performance issues. Do them separately.
+  while (1) {
+    @autoreleasepool {
+      if (breakDir || progress.isCancelled) break;
+
+      // Wait for a processing thread to become available
+      dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+
+      dispatch_group_async(group,
+                           dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+        pthread_mutex_lock(&enumeratorMutex);
+        NSString *file = [dirEnum nextObject];
+        fileCount++;
+        pthread_mutex_unlock(&enumeratorMutex);
+
+        if (!file) {
+          breakDir = YES;
+          dispatch_semaphore_signal(sema);
+          return;
+        }
+
+        if ([dirEnum fileAttributes][NSFileType] != NSFileTypeRegular) {
+          dispatch_semaphore_signal(sema);
+          return;
+        }
+
+        NSString *newFile = [event.fileBundlePath stringByAppendingPathComponent:file];
+        SNTFileInfo *fi = [[SNTFileInfo alloc] initWithPath:newFile];
+        if (!fi.isExecutable) {
+          dispatch_semaphore_signal(sema);
+          return;
+        }
+
+        pthread_mutex_lock(&eventsMutex);
+        [fis addObject:fi];
+        binaryCount++;
+        pthread_mutex_unlock(&eventsMutex);
+
+        dispatch_semaphore_signal(sema);
+      });
+      if (progress && ((fileCount % 500) == 0 || binaryCount > sentBinaryCount)) {
+        sentBinaryCount = binaryCount;
+        [[self.notifierConnection remoteObjectProxy] updateTotalFileCountForEvent:event
+                                                                        binaryCount:binaryCount
+                                                                        fileCount:fileCount];
+      }
+    }
+  }
+
+  dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+
+  NSProgress *p;
+  if (progress) {
+    [progress becomeCurrentWithPendingUnitCount:1];
+    p = [NSProgress progressWithTotalUnitCount:fis.count];
+  }
+
+  // Perform the hashing of found binaries in this stage.
+  for (SNTFileInfo *fi in fis) {
+    @autoreleasepool {
+      if (progress.isCancelled) break;
+
+      // Wait for a processing thread to become available
+      dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+
+      dispatch_group_async(group,
+                           dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+        @autoreleasepool {
+          SNTStoredEvent *se = [[SNTStoredEvent alloc] init];
+          se.filePath = fi.path;
+          se.fileSHA256 = fi.SHA256;
+          se.occurrenceDate = [NSDate distantFuture];
```

Comment:
Unfortunately not. We require an occurrenceDate to be stored in the events database. This is striped out before uploading to the sync server.